app/validate_txs.go

```diff
@@ -8,11 +8,44 @@ import (
 	"github.com/cosmos/cosmos-sdk/telemetry"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
+	square "github.com/celestiaorg/go-square/v2"
 	"github.com/celestiaorg/go-square/v2/tx"
 
 	"github.com/celestiaorg/celestia-app/v4/pkg/appconsts"
 )
 
+// FilteredBuilder filters txs and blobs using a copy of the state and tx validity
+// rules before adding it the square.
+type FilteredBuilder struct {
+	logger   log.Logger
+	ctx      sdk.Context
+	handler  sdk.AnteHandler
+	txConfig client.TxConfig
+	builder  *square.Builder
+}
+
+func NewFilteredBuilder(
+	logger log.Logger,
+	ctx sdk.Context,
+	handler sdk.AnteHandler,
+	txConfig client.TxConfig,
+	maxSquareSize,
+	subtreeRootThreshold int,
+) (*FilteredBuilder, error) {
+	builder, err := square.NewBuilder(maxSquareSize, subtreeRootThreshold)
```

Comment:
[nit] it can be tricky to debug functions that return a non nil value AND an error. IMO for new code, we should return value or error but never both.

```suggestion
	builder, err := square.NewBuilder(maxSquareSize, subtreeRootThreshold)
	if err != nil {
		return nil, err
	}
```